src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+        {
+            if (username == null && (password != null || sessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (username != null && password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
+        }
+
+        private static void ValidateCredentials(string username, SecureString password, string sessionToken)
+        {
+            if (username == null && (password != null || sessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (username != null && password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
+        }
+
+        // nested classes
+        private class AwsCredentials
+        {
+            public UsernamePasswordCredential Credentials;
+            public string SessionToken;
+        }
+
+        private class MongoAWSMechanism : ISaslMechanism
+        {
+            private readonly UsernamePasswordCredential _credential;
+            private readonly string _name;
+            private readonly IRandomByteGenerator _randomByteGenerator;
+            private readonly string _sessionToken;
+
+            public MongoAWSMechanism(
+                UsernamePasswordCredential credential,
+                string sessionToken,
+                IRandomByteGenerator randomByteGenerator)
+            {
+                _name = "MONGODB-AWS";
+                _credential = Ensure.IsNotNull(credential, nameof(credential));
+                _sessionToken = sessionToken;
+                _randomByteGenerator = Ensure.IsNotNull(randomByteGenerator, nameof(randomByteGenerator));
+            }
+
+            public string Name
+            {
+                get { return _name; }
+            }
+
+            public ISaslStep Initialize(IConnection connection, SaslConversation conversation, ConnectionDescription description)
+            {
+                Ensure.IsNotNull(connection, nameof(connection));
+                Ensure.IsNotNull(description, nameof(description));
+
+                var nonce = GenerateRandomBytes();
+
+                var document = new BsonDocument()
+                    .Add("r", new BsonBinaryData(nonce))
+                    .Add("p", new BsonInt32('n'));
+
+                var clientFirstMessageBytes = ToBytes(document);
+
+                return new ClientFirst(clientFirstMessageBytes, nonce, _credential, _sessionToken);
+            }
+
+            private byte[] GenerateRandomBytes()
+            {
+                return _randomByteGenerator.Generate(randomLength);
+            }
+        }
+
+        private static BsonDocument ToDocument(byte[] bytes)
```

Comment:
I'm not sure that this is the best way to do this work. However maybe I missed something, I need to understand what data are presented in bytes to be sure. I think we need to have a test in Core with mocked data similar to: https://github.com/mongodb/mongo-csharp-driver/blob/master/tests/MongoDB.Driver.Core.Tests/Core/Authentication/ScramSha256AuthenticatorTests.cs#L33. This will allow us to be able to run these code paths without any additional configuration